### ISSUE TITLE
Unittest test_no_auth() fails (Branch 15c)

### ISSUE THREAD

**Issue by ScottMcCormack** [miguelgrinberg/flasky#83]
test_no_auth() (from testing/test_api.py) fails when using running unit testing on branch 15c

In this test case response.status_code returns a code of 200 not 401.

**Comment by miguelgrinberg** [comment]
I cannot reproduce this test failure. Do you have anything different in your checked out copy?

    (venv) flasky (6e27193*) $ ./manage.py test
    test_404 (test_api.APITestCase) ... ok
    test_anonymous (test_api.APITestCase) ... ok
    test_bad_auth (test_api.APITestCase) ... ok
    test_comments (test_api.APITestCase) ... ok
    test_no_auth (test_api.APITestCase) ... ok
    test_posts (test_api.APITestCase) ... ok
    test_token_auth (test_api.APITestCase) ... ok
    test_users (test_api.APITestCase) ... ok
    test_app_exists (test_basics.BasicsTestCase) ... ok
    test_app_is_testing (test_basics.BasicsTestCase) ... ok
    test_home_page (test_client.FlaskClientTestCase) ... ok
    test_register_and_login (test_client.FlaskClientTestCase) ... ok
    test_anonymous_user (test_user_model.UserModelTestCase) ... ok
    test_duplicate_email_change_token (test_user_model.UserModelTestCase) ... ok
    test_expired_confirmation_token (test_user_model.UserModelTestCase) ... ok
    test_follows (test_user_model.UserModelTestCase) ... ok
    test_gravatar (test_user_model.UserModelTestCase) ... ok
    test_invalid_confirmation_token (test_user_model.UserModelTestCase) ... ok
    test_invalid_email_change_token (test_user_model.UserModelTestCase) ... ok
    test_invalid_reset_token (test_user_model.UserModelTestCase) ... ok
    test_no_password_getter (test_user_model.UserModelTestCase) ... ok
    test_password_salts_are_random (test_user_model.UserModelTestCase) ... ok
    test_password_setter (test_user_model.UserModelTestCase) ... ok
    test_password_verification (test_user_model.UserModelTestCase) ... ok
    test_ping (test_user_model.UserModelTestCase) ... ok
    test_roles_and_permissions (test_user_model.UserModelTestCase) ... ok
    test_timestamps (test_user_model.UserModelTestCase) ... ok
    test_valid_confirmation_token (test_user_model.UserModelTestCase) ... ok
    test_valid_email_change_token (test_user_model.UserModelTestCase) ... ok
    test_valid_reset_token (test_user_model.UserModelTestCase) ... ok
    
    ----------------------------------------------------------------------
    Ran 30 tests in 5.667s
    
    OK

**Comment by DoctorRudi** [comment]
Hi Minguel,

My first mistake was that I was trying to run my tests in my own python environment, not the venv with the correct files specified by the requirements\common.txt file.

I've done a bit of testing and it looks as though the problem that I have had is with the version of Flask-HTTPAuth that I was using. I was using Flask-HTTPAuth==2.7.0 instead of Flask-HTTPAuth==2.2.0 as specified in requirements\common.txt. In this case

```python
self.assertTrue(response.status_code ==  401)
```
Returns a code of 200 instead of 401

Below is the log that I get when I run the tests with Flask-HTTPAuth==2.7.0 installed instead of Flask-HTTPAuth==2.2.0.

```
test_404 (test_api.APITestCase) ... ok
test_anonymous (test_api.APITestCase) ... ok
test_bad_auth (test_api.APITestCase) ... ok
test_comments (test_api.APITestCase) ... ok
test_no_auth (test_api.APITestCase) ... FAIL
test_posts (test_api.APITestCase) ... ok
test_token_auth (test_api.APITestCase) ... ok
test_unconfirmed_account (test_api.APITestCase) ... ok
test_users (test_api.APITestCase) ... ok
test_app_exists (test_basics.BasicsTestCase) ... ok
test_app_is_testing (test_basics.BasicsTestCase) ... ok
test_home_page (test_client.FlaskClientTestCase) ... ok
test_register_and_login (test_client.FlaskClientTestCase) ... ok
test_anonymous_user (test_user_model.UserModelTestCase) ... ok
test_duplicate_email_change_token (test_user_model.UserModelTestCase) ... ok
test_expired_confirmation_token (test_user_model.UserModelTestCase) ... ok
test_follows (test_user_model.UserModelTestCase) ... ok
test_gravatar (test_user_model.UserModelTestCase) ... ok
test_invalid_confirmation_token (test_user_model.UserModelTestCase) ... ok
test_invalid_email_change_token (test_user_model.UserModelTestCase) ... ok
test_invalid_reset_token (test_user_model.UserModelTestCase) ... ok
test_no_password_getter (test_user_model.UserModelTestCase) ... ok
test_password_salts_are_random (test_user_model.UserModelTestCase) ... ok
test_password_setter (test_user_model.UserModelTestCase) ... ok
test_password_verification (test_user_model.UserModelTestCase) ... ok
test_ping (test_user_model.UserModelTestCase) ... ok
test_roles_and_permissions (test_user_model.UserModelTestCase) ... ok
test_timestamps (test_user_model.UserModelTestCase) ... ok
test_valid_confirmation_token (test_user_model.UserModelTestCase) ... ok
test_valid_email_change_token (test_user_model.UserModelTestCase) ... ok
test_valid_reset_token (test_user_model.UserModelTestCase) ... ok

======================================================================
FAIL: test_no_auth (test_api.APITestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "Z:\Documents\GitHub\flasky\tests\test_api.py", line 42, in test_no_auth
    self.assertTrue(response.status_code == 401)
AssertionError: False is not true

----------------------------------------------------------------------
Ran 31 tests in 40.587s
```

**Comment by Helisquinde** [comment]
I have exactly the same problem. When I use Flask-HTTPAuth==2.7.0 instead of 2.2.0, test_no_auth fails, with the same error log as @DoctorRudi

**Comment by alejoar** [comment]
Can confirm this error, but it is not surprising. I thought the posts collection was intentionally available with no auth, so I would say the test is incorrect.

Testing with HTTPie:
```
(venv) $ http --json GET http://localhost:5000/api/v1.0/posts/
HTTP/1.0 200 OK
Content-Length: 8699
Content-Type: application/json
Date: Thu, 29 Oct 2015 19:03:17 GMT
Server: Werkzeug/0.10.4 Python/2.7.6
```
I get the posts collection correctly even without using --auth. Same for individual post resource (/posts/x).

So, there are two solutions:
a) Change the test so that it asserts an HTTP 200 response:
```python
self.assertTrue(response.status_code == 401)
```
b) Change the permission access for posts. As there is no "READ_POSTS" permission, it needs to be added to the Permission class (models.py):
```python
class Permission:
    FOLLOW = 0x01               # 0000 0001
    COMMENT = 0x02              # 0000 0010
    WRITE_ARTICLES = 0x04       # 0000 0100
    MODERATE_COMMENTS = 0x08    # 0000 1000
    READ_POSTS = 0x10           # 0001 0000
    ADMINISTER = 0x80           # 1000 0000
```
Users and moderators need this permission too:
```python
class Role(db.Model):
    # ...
    @staticmethod
    def insert_roles():
        roles = {
            'User': (Permission.FOLLOW |
                     Permission.COMMENT |
                     Permission.WRITE_ARTICLES |
                     Permission.READ_POSTS, True),
            'Moderator': (Permission.FOLLOW |
                          Permission.COMMENT |
                          Permission.WRITE_ARTICLES |
                          Permission.READ_POSTS |
                          Permission.MODERATE_COMMENTS, False),
            'Administrator': (0xff, False)
        }
    # ...
```
Finally, in the api:
```python
@api.route('/posts/')
@permission_required(Permission.READ_POSTS)
def get_posts():
    # ...
```
and:
```python
@api.route('/posts/<int:id>')
@permission_required(Permission.READ_POSTS)
def get_post(id):
    # ...
```

The correct solution is solution **a**, as it doesn't make sense to restrict the posts resource for anonymous users given that they can see the posts in the web version.

Also, if **b** is chosen as a solution, the test *test_anonymous* needs to be changed to verify response is 403 and not 200, as now viewing posts would be a privileged permission not accessible to anonymous users:
```python
def test_anonymous(self):
    response = self.client.get(
        url_for('api.get_posts'),
        headers=self.get_api_headers('', ''))
    self.assertTrue(response.status_code == 403)
```

Hope this helps @Helisquinde and @ScottMcCormack understand why you are getting the error. As for the Flask-HTTPAuth version, I am using 2.7.0 and didn't test with any previous version, but as I explained, and as long as I'm not mistaken, the behavior is normal and there is just a mistake in the test.

**Comment by miguelgrinberg** [comment]
This problem is now fixed. There were actually two related problems combined, and that caused part of the confusion.

The first issue is that, as @alejoar indicates, the API is designed to accept anonymous users. Anybody can request posts, without providing authentication. So this unit test should ensure that the status code returned by the GET request to /posts is 200, not 401.

The reason why with Flask-HTTPAuth 2.2.0 the test passes is that in that release of the extension anonymous requests were rejected without consulting with the application. This was corrected in release 2.4.0, at which point anonymous requests are also passed through the `verify_password` callback.
